packages/paima-sdk/paima-events/src/app-events.ts

```diff
@@ -0,0 +1,200 @@
+> & {
+  get: () => {
+    address: `0x${string}`;
+    data: { name: string; fields: KeypairToObj<T[number]['fields']>; topic: string };
+  }[];
+};
+
+// Typed stateful builder that handles overloaded topics. There is a performance
+// cost of using this, since it dynamically finds the matching overloaded topic
+// from the provided fields, but this way there is no need to use indexing to
+// provide the right event type. If this is a concern then the other helper
+// function can be used to explicitly provide the event type.
+//
+// This is supposed meant to be used like this (pseudo-code):
+//
+// const eventsDefinitions = [
+//   genEvent({ name: 'A', fields: fields1 }),
+//   genEvent({ name: 'B', fields: fields2 }),
+//   genEvent({ name: 'B', fields: fields3 }),
+// ] as const;
+//
+// const queueFactory = eventQueueFactory(eventsDefinitions);
+// const queue = queueFactory.empty();
+//
+// queue.emit('A', data1);
+// queue.emit('B', data2);
+// queue.emit('B', data3);
+//
+// return { events: queue, stateTransitions: {...}};
+//
+// Where data2 and data3 may be different types (for the two overloaded
+// definitions), and the return type is that of the stf.
+export function eventQueueFactory<
+  const T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>,
+>(arr: readonly [...T]): { empty: () => EventQueue<T> } {
+  const prototype: Partial<EmitFunction<T, T[number]['name']>> = {};
+
+  const topicSearch = {} as Record<
+    T[number]['name'],
+    { topic: string; check: (fields: UnionForOverloadedEvents<T, string>) => boolean }[]
+  >;
+
+  arr.forEach((type: T[number]) => {
+    const name: T[number]['name'] = type.name;
+    if (!topicSearch[name]) {
+      topicSearch[name] = [];
+    }
+
+    const topics = topicSearch[name];
+
+    topics.push({
+      topic: toTopicHash(type),
+      check: (fields: UnionForOverloadedEvents<T, string>) => {
+        const fieldsMatch = type.fields.every(definition => {
+          const value = fields[definition.name];
+
+          return value && Value.Check(definition.type, value);
+        });
+
+        const noExtraFields = Object.keys(fields).length === type.fields.length;
+
+        return fieldsMatch && noExtraFields;
+      },
+    });
+  });
+
+  prototype.emit = function (name, address, fields): void {
+    const topics = topicSearch[name];
+
+    let topic;
+    // Go backwards to find the 'latest' overload, assuming that new events
+    // get always added at the end. Generally there shouldn't be more than one
+    // match, but it's technically possible if the types are unions.
+    for (let i = topics.length - 1; i >= 0; i--) {
+      if (topics[i].check(fields)) {
+        topic = topics[i].topic;
+        break;
+      }
+    }
+
+    if (!topic) {
+      // this shouldn't really happen because of the static typing, but it
+      // can't be inferred.
+      // it technically can also happen if there are extra fields, since
+      // typescript doesn't guarantee that.
```

Comment:
actually, I thought this was the case? But it doesn't seem to happen now that I tested it.